simple/rdm_triggeredOps.c

```diff
@@ -0,0 +1,670 @@
+	iov->iov_len = size;
+	msg->msg_iov = iov;
+}
+
+static void format_simple_msg_rma(struct fi_msg_rma *msg, struct iovec *iov,
+			   struct fi_rma_iov *rma_iov, void *src,
+			   size_t size, void *ctx)
+{
+	msg->context = ctx;
+	msg->desc = fi_mr_desc(mr);
+	msg->iov_count = 1;
+	msg->addr = remote_fi_addr;
+	msg->rma_iov_count = 1;
+
+	iov->iov_base = src;
+	iov->iov_len = size;
+	msg->msg_iov = iov;
+
+	rma_iov->addr = 0;
+	rma_iov->key = FT_MR_KEY;
+	rma_iov->len = size;
+	msg->rma_iov = rma_iov;
+}
+
+static void format_simple_msg_atomic(struct fi_msg_atomic *msg, struct fi_ioc *iov,
+			      struct fi_rma_ioc *rma_iov, void *src, size_t size,
+			      void *ctx, enum fi_datatype datatype,
+			      enum fi_op op)
+{
+	msg->context = ctx;
+	msg->desc = fi_mr_desc(mr);
+	msg->iov_count = 1;
+	msg->rma_iov_count = 1;
+	msg->addr = remote_fi_addr;
+	msg->data = 0;
+	msg->datatype = datatype;
+	msg->op = op;
+	msg->data = 0;
+
+	iov->addr = src;
+	iov->count = size;
+	msg->msg_iov = iov;
+
+	rma_iov->addr = 0;
+	rma_iov->count = size;
+	rma_iov->key = FT_MR_KEY;
+	msg->rma_iov = rma_iov;
+}
+
+static void format_simple_msg_fetch(struct fi_msg_fetch *msg, struct fi_ioc *iov,
+				    void *src, size_t size)
+{
+	msg->desc = fi_mr_desc(mr_result);
+
+	iov->addr = src;
+	iov->count = size;
+
+	msg->iov_count = 1;
+	msg->msg_iov = iov;
+}
+
+static void format_simple_msg_compare(struct fi_msg_compare *msg, struct fi_ioc *iov,
+				      void *src, size_t size)
+{
+	msg->desc = fi_mr_desc(mr_compare);
+
+	iov->addr = src;
+	iov->count = size;
+
+	msg->iov_count = 1;
+	msg->msg_iov = iov;
+}
+
+static int check_data()
+{
+	int ret, i;
+	char c = ~0;
+
+	switch (tested_op) {
+	case FI_OP_COMPARE_ATOMIC:
+	case FI_OP_FETCH_ATOMIC:
+		if (strcmp(result_buf, welcome_text)) {
+			printf("Data mismatch in fetch buffer, found: %s",
+				result_buf);
+			return 1;
+		}
+	case FI_OP_ATOMIC:
+		for (i = 0; i < strlen(welcome_text); i++) {
+			if (rx_buf[i] != c) {
+				printf("Data mismatch found at byte %d", i);
+				return 1;
+			}
+		}
+		break;
+	case FI_OP_CNTR_SET:
+	case FI_OP_CNTR_ADD:
+		ret  = fi_cntr_wait(test_cntr, 10, -1);
+		if (ret)
+			return ret;
+		break;
+	case FI_OP_READ:
+	case FI_OP_WRITE:
+		if (strcmp(result_buf, write_text)) {
+			printf("Data mismatch in read buffer, found: %s",
+				result_buf);
+			return 1;
+		}
+	default:
+		if (strcmp(rx_buf, write_text)) {
+			printf("Data mismatch in rx_buf, found: %s", rx_buf);
+			return 1;
+		}
+	}
+	return 0;
+}
+
+static int send_wait_check()
+{
+	int ret;
+
+	if (opts.dst_addr) {
+		ret = fi_write(ep, tx_buf, strlen(welcome_text), fi_mr_desc(mr),
+				remote_fi_addr, 0, FT_MR_KEY, &tx_ctx);
+ 		if (ret) {
+ 			FT_PRINTERR("fi_write", ret);
+ 			return ret;
+		}
+	}
+
+	ret = fi_cntr_wait(test_cntr, n_trig, -1);
+	if (ret)
+		return ret;
+
+	ret = fi_write(ep, tx_buf, strlen(write_text), fi_mr_desc(mr),
+			remote_fi_addr, strlen(write_text) + 1, FT_MR_KEY, &tx_ctx);
```

Comment:
Ahh, that was before the receive completion revelation. The first write is the trigger. The second I added in to sync up both sides before the verification but that shouldn't be needed now that I understand the completion generation. I'll take that out.